test/jdk/java/foreign/TestUpcallException.java

```diff
@@ -37,22 +37,126 @@
 import org.testng.annotations.Test;
 
 import java.io.IOException;
+import java.lang.foreign.Arena;
+import java.lang.foreign.FunctionDescriptor;
+import java.lang.foreign.Linker;
+import java.lang.foreign.MemorySegment;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
 
 public class TestUpcallException extends UpcallTestHelper {
 
-    @Test(dataProvider = "cases")
-    public void testException(boolean useSpec, boolean isVoid) throws InterruptedException, IOException {
-        runInNewProcess(ThrowingUpcall.class, useSpec, isVoid ? "void" : "")
+    @Test(dataProvider = "exceptionCases")
+    public void testException(Class<?> target, boolean useSpec) throws InterruptedException, IOException {
+        runInNewProcess(target, useSpec)
                 .assertStdErrContains("Testing upcall exceptions");
     }
 
     @DataProvider
-    public static Object[][] cases() {
+    public static Object[][] exceptionCases() {
         return new Object[][]{
-            { false, true,  },
-            { false, false, },
-            { true,  true,  },
-            { true,  false, }
+            { VoidUpcallRunner.class,    false },
+            { NonVoidUpcallRunner.class, false },
+            { VoidUpcallRunner.class,    true  },
+            { NonVoidUpcallRunner.class, true  }
         };
     }
+
+    public static class VoidUpcallRunner extends ExceptionRunnerBase {
+        public static void main(String[] args) throws Throwable {
+            try (Arena arena = Arena.openConfined()) {
+                MemorySegment stub = Linker.nativeLinker().upcallStub(VOID_TARGET, FunctionDescriptor.ofVoid(), arena.scope());
+
+                downcallVoid.invoke(stub); // should call Shutdown.exit(1);
+            }
+        }
+    }
+
+    public static class NonVoidUpcallRunner extends ExceptionRunnerBase {
+        public static void main(String[] args) throws Throwable {
+            MethodHandle handle = MethodHandles.identity(int.class);
```

Comment:
I'm not 100% sure that building these complex method handle chains is better than just having two throwing methods in the code - one with zero arguments and one with one argument (int) ?

Reply:
Yeah, that seems simpler.